File: 20/20.py


class Solution(Object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stk = [];
        for c in s:
            if c in ['(', '{', '[']:
                stk.append(c)
            else:
                if len(stk) != 0:
                    cinstk = stk.pop()
                    if cinstk == '(' and c != ')':
                        return False
                    elif cinstk == '{' and c != '}':
                        return False
                    elif cinstk == '[' and c != ']':
                        return False
                else:
                    return False

        return len(stk) == 0

    # nice code
    def isValid2(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        dict = {"]":"[","}":"{",")":"("}
        for c in s:
            if c in dict.values():
                stack.append(c)
            elif c in dict.keys():
                if stack == [] or dict[c] != stack.pop():
                    return False
            else:
                return False

        return stack == []

File: backup/20/20.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
    public:
        bool isValid(string s) {
            stack<char> st;
            string::iterator it;
            for (it = s.begin(); it != s.end(); it++) {
                if(*it == '(' || *it == '{' || *it == '[') 
                    st.push(*it);
                else {
                    if (!st.empty()) {
                        char c = st.top();
                        st.pop();
                        if( c == '(' && *it != ')') return false;
                        else if (c == '{' && *it != '}') return false;
                        else if (c == '[' && *it != ']') return false;
                    }
                    else {
                        return false;
                    }
                }

            }
            return st.empty();
        }
};
